buttercup/cogs/history.py

```diff
@@ -0,0 +1,284 @@
+    """Add entries for the zero rates to the data frame.
+
+    When the rate is zero, it is not returned in the API response.
+    Therefore we need to add it manually.
+    However, for a span of zero entries, we only need the first
+    and last entry. This reduces the number of data points.
+    """
+    new_index = set()
+    delta = get_timedelta_from_time_frame(time_frame)
+    now = datetime.now(tz=tzutc())
+
+    for date in data.index:
+        new_index.add(date)
+        new_index.add(date - delta)
+        if date + delta < now:
+            new_index.add(date + delta)
+
+    return data.reindex(new_index, fill_value=0).sort_index()
+
+
+def add_rank_lines(ax: plt.Axes, gamma: int) -> plt.Axes:
+    """Add the rank lines to the given axes."""
+    # Show ranks when you are close to them already
+    threshold_factor = 0.7
+    for rank in ranks:
+        if gamma >= rank["threshold"] * threshold_factor:
+            ax.axhline(y=rank["threshold"], color=rank["color"], zorder=-1)
+    return ax
+
+
+def create_file_from_figure(fig: plt.Figure, file_name: str) -> File:
+    """Create a Discord file containing the figure."""
+    history_plot = io.BytesIO()
+
+    fig.savefig(history_plot, format="png")
+    history_plot.seek(0)
+    plt.close(fig)
+
+    return File(history_plot, file_name)
+
+
+class History(Cog):
+    def __init__(self, bot: ButtercupBot, blossom_api: BlossomAPI) -> None:
+        """Initialize the History cog."""
+        self.bot = bot
+        self.blossom_api = blossom_api
+
+    def get_user_history(self, user: str) -> Tuple[int, pd.DataFrame]:
+        """Get a data frame representing the history of the user.
+
+        :returns: The gamma of the user and their history data.
+        """
+        page_size = 500
+
+        # First, get the total gamma for the user
+        user_response = self.blossom_api.get_user(user)
+        if user_response.status != BlossomStatus.ok:
+            raise BlossomException(user_response)
+
+        user_gamma = user_response.data["gamma"]
+        user_id = user_response.data["id"]
+
+        time_frame = get_data_granularity(user_gamma)
+
+        # Get all rate data
+        user_data = pd.DataFrame(columns=["date", "count"]).set_index("date")
+        page = 1
+        response = self.blossom_api.get(
+            f"volunteer/{user_id}/rate",
+            params={"page": page, "page_size": page_size, "time_frame": time_frame},
+        )
+
+        while response.status_code == 200:
+            rate_data = response.json()["results"]
+
+            new_frame = pd.DataFrame.from_records(rate_data)
+            # Convert date strings to datetime objects
+            new_frame["date"] = new_frame["date"].apply(lambda x: parser.parse(x))
+            # Add the data to the list
+            user_data = user_data.append(new_frame.set_index("date"))
+
+            # Continue with the next page
+            page += 1
+            response = self.blossom_api.get(

```

Comment:
We can use the "next" URL in combination with a while loop where we start with the URL used in lines `134-135`. Using this `next`, we can get rid of `158-160`, which seems like a workaround which is in this case not necessary.

Reply:
Can we easily drop in a full URL in the request? I was thinking the `get` method of the wrapper omitted the base URL, so I wasn't sure how to do it cleanly.

Reply:
You are indeed right that the wrapper uses the base URL. However, I think we can atleast check `next` whether it is non-empty rather than relying on the 404

Reply:
That's true, I fixed that!